spec.md

```diff
@@ -0,0 +1,276 @@
+
+### PUT `/ipns/<key>[/<path>]`
+
+The same as POST `/ipns/<key>[/<path>]`, but where the response body is the file to be uploaded, not an IPFS path. Multiple files can be uploaded using form data, in which case files will go inside the directory in the path.
+
+### PUT `/ipns/localhost[/<path>]`
+
+The same as POST `/ipns/localhost[/<path>]`, but where the response body is the file to be uploaded, not an IPFS path. Multiple files can be uploaded using form data, in which case files will go inside the directory in the path.
+
+### DELETE `/ipns/<key>[/<path>]`
+
+Remove content behind an IPNS name. The new content is pinned, and the previous content is unpinned.
+
+#### Response
+A 301 Temporary Redirect to `ipns://<key>[/<path>]`. The path in the redirect is the directory containing the deletion path. So if the path in the URL was `/ipns/<key>/path/to/file.ext`, the redirect will be to `ipns://<key>/path/to`.
+
+The header `X-IPNS-Path` is set to `/ipns/<key>[/<path>]`, with the same path logic as above.
+
+### DELETE `/ipns/localhost[/<path>]`
+
+Behaves the same as `/ipns/<key>[/<path>]`.
+
+#### Query params
+- `key=some_name`: **Required**. This specifies the key you want by a name instead of by public key. If a key with this name doesn't exist, it will be created.
+
+#### Response
+The same as `/ipns/<key>[/<path>]`. Note this allows for extraction of the actual public key from future use and link sharing.
+
+## Pubsub
+
+An experimental pubsub API is supported to give users direct access to sending and receiving pubsub messages.
+
+### GET `/pubsub/<topic>`
+
+#### Query params
+- `format`: can be `json`, `utf8`, `base64`, or not specified, which defaults to `base64`. This controls the decoding of pubsub message bytes
+
+#### Response
+
+The response is an [event stream](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format), so pubsub messages for that topic will be sent to the client as long as the connection remains open.
+
+Pubsub messages are encoded as events like this:
+
+```
+id: FuhB2orEIwE=
+data: {"from":"QmfULfLaRtF7Kg3ShjFdiYL97P7eEcyL4PcvQLhkzCPCeG","data":"aGVsbG8=","topics":["test"]}
+```
+
+The "message identifier" of the pubsub message is base64-encoded to form the `id:`. `data:` is always encoded as JSON, with the other three things known about a pubsub message.
+
+In the JSON: `from` is the prettified ID of the node who sent the message, `data` is the message data, and `topics` is an array of topics this message was sent to.
+
+In the JSON, `data` is either a JSON object, a string, or bytes base64-encoded into a string. This behaviour depends on the format parameter mentioned above.
+
+The response also contains the header `X-IPFS-ID` which is this node's prettified ID. This allows the client to filter out pubsub messages that come from itself.
```

Comment:
I don't think this can be reached via the [EventSouce](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) API.

Would we maybe need to put it into the `HEAD` request if we don't do the `GET {Accept: application/json}` thing?

Reply:
Hmm that's too bad. But yeah I'll put it in HEAD. It can also go in GET as you mention, if that ends up being implemented.

Reply:
Added to code in 212017c and to spec in 51b8886

Reply:
Since we're doing the JSON thing, might as well remove it from HEAD

Reply:
I think it's okay to leave HEAD in, and might even be required for CORS. There's no maintenance cost. Supporting HEAD matches the behaviour of other protocols like `ipfs` and `ipns` that support it.

Reply:
Like, leave the header in HEAD?

I was thinking we could get rid of that extra header entirely since it's in the JSON response. Or is there other uses for it?

Reply:
Removed the header in 4888f68